lib/mix/lib/mix/tasks/compile.all.ex

```diff
@@ -15,9 +15,13 @@ defmodule Mix.Tasks.Compile.All do
 
     config = Mix.Project.config()
 
-    Mix.Project.with_build_lock(config, fn ->
+    if "--no-compile" in args do
       do_run(config, args)
-    end)
+    else
+      Mix.Project.with_build_lock(config, fn ->
```

Comment:
This is an interesting change. Although `do_run` above does not compile any application, it still tries to load them. So there is a chance that we try to load the applications while someone is compiling, and therefore erasing modules and .app files. Or someone could be running `clean`. I wonder if we should just always keep the lock? If someone cleans, it will fail anyway, but at least it will do so more consistently and not load an intermediate state? (and then I also wonder if we should have a deps lock while loading deps apps).